tensorflow/contrib/rnn/python/ops/rnn_cell.py

```diff
@@ -3392,3 +3392,273 @@ def call(self, inputs, state):
+      w_list.append(w)
+
+    # Reading (Sec 3.1)
+
+    read_w_list = w_list[:self.read_head_num]
+    read_vector_list = []
+    for i in range(self.read_head_num):
+      read_vector = math_ops.reduce_sum(
+        array_ops.expand_dims(read_w_list[i], dim=2) * prev_M, axis=1)
+      read_vector_list.append(read_vector)
+
+    # Writing (Sec 3.2)
+
+    write_w_list = w_list[self.read_head_num:]
+    M = prev_M
+    for i in range(self.write_head_num):
+      w = array_ops.expand_dims(write_w_list[i], axis=2)
+      erase_vector = array_ops.expand_dims(
+        math_ops.sigmoid(erase_add_list[i * 2]), axis=1)
+      add_vector = array_ops.expand_dims(
+        math_ops.tanh(erase_add_list[i * 2 + 1]), axis=1)
+      erase_M = array_ops.ones_like(M) - math_ops.matmul(w, erase_vector)
+      M = M * erase_M + math_ops.matmul(w, add_vector)
+
+    output = math_ops.matmul(
+      array_ops.concat([controller_output] + read_vector_list, axis=1),
+      self._output_kernel)
+    output = nn_ops.bias_add(output, self._output_bias)
+    output = clip_ops.clip_by_value(output,
+      -self.clip_value, self.clip_value)
+
+    return output, NTMControllerState(
+      controller_state=controller_state, read_vector_list=read_vector_list,
+      w_list=w_list, M=M)
+
+  def addressing(self, k, beta, g, s, gamma, prev_M, prev_w):
+    # Sec 3.3.1 Focusing by Content
+
+    k = array_ops.expand_dims(k, axis=2)
+    inner_product = math_ops.matmul(prev_M, k)
+    k_norm = math_ops.sqrt(math_ops.reduce_sum(
+      math_ops.square(k), axis=1, keepdims=True))
+    M_norm = math_ops.sqrt(math_ops.reduce_sum(
+      math_ops.square(prev_M), axis=2, keepdims=True))
+    norm_product = M_norm * k_norm
+    
+    # eq (6)
+    K = array_ops.squeeze(inner_product / (norm_product + 1e-8))
+
+    K_amplified = math_ops.exp(array_ops.expand_dims(beta, axis=1) * K)
+    
+    # eq (5)
+    w_c = K_amplified / math_ops.reduce_sum(K_amplified, axis=1, keepdims=True)
+
+    # Sec 3.3.2 Focusing by Location
+
+    g = array_ops.expand_dims(g, axis=1)
+    
+    # eq (7)
+    w_g = g * w_c + (1 - g) * prev_w
+
+    s = array_ops.concat([s[:, :self.shift_range + 1],
+      array_ops.zeros([s.get_shape()[0],
+        self.memory_size - (self.shift_range * 2 + 1)]),
+      s[:, -self.shift_range:]], axis=1)
+    t = array_ops.concat([array_ops.reverse(s, axis=[1]),
+      array_ops.reverse(s, axis=[1])], axis=1)
+    s_matrix = array_ops.stack(
+      [t[:, self.memory_size - i - 1:self.memory_size * 2 - i - 1]
+        for i in range(self.memory_size)],
+      axis=1)
+    
+    # eq (8)
+    w_ = math_ops.reduce_sum(
+      array_ops.expand_dims(w_g, axis=1) * s_matrix, axis=2)
+    w_sharpen = math_ops.pow(w_, array_ops.expand_dims(gamma, axis=1))
+
+    # eq (9)
+    w = w_sharpen / math_ops.reduce_sum(w_sharpen, axis=1, keepdims=True)
+
+    return w
+
+  def zero_state(self, batch_size, dtype):
+    self._init_read_vectors = [self.add_variable(
```

Comment:
is there a way you could move all the variable creation to build() and access it in call() instead of creating a new variable each time zero_state is called?

the current code will be pretty problematic if you use the same cell in two calls to dynamic_rnn; because dynamic_rnn calls zero_state, so for example some of these variables will have already been created and it may cause an error.

in contrast, if you move all variable building into build(), the initial vector variables will simply be reused across the calls.

Reply:
(similarly zero_state should be functional; so setting self._ properties inside it is a really bad idea)

Reply:
Yes, so I had tried that previously but get the following error about the variables not having been created when the zero_state method is called:

```
Traceback (most recent call last):
  File "run_tasks.py", line 145, in <module>
    train_model = BuildTrainModel(train_max_seq_len, train_inputs, train_outputs)
  File "run_tasks.py", line 119, in __init__
    super(BuildTrainModel, self).__init__(max_seq_len, inputs, tf.contrib.learn.ModeKeys.TRAIN)
  File "run_tasks.py", line 67, in __init__
    self._build_model()
  File "run_tasks.py", line 107, in _build_model
    initial_state=initial_state if args.mann == 'none' else None)
  File "/Users/mcollier/anaconda2/lib/python2.7/site-packages/tensorflow/python/ops/rnn.py", line 601, in dynamic_rnn
    state = cell.zero_state(batch_size, dtype)
  File "/Users/mcollier/anaconda2/lib/python2.7/site-packages/tensorflow/contrib/rnn/python/ops/rnn_cell.py", line 3305, in zero_state
    for i in range(self.read_head_num)]
AttributeError: 'NTMCell' object has no attribute '_init_read_vectors'
```

Perhaps I am doing something wrong?